README.md
# GreaterLess

[![Build Status](https://travis-ci.org/roqua/greater_less.svg)](https://travis-ci.org/roqua/greater_less)
[![Dependency Status](https://gemnasium.com/badges/github.com/roqua/greater_less.svg)](https://gemnasium.com/github.com/roqua/greater_less)

The GreaterLess gem can be used to generate objects that represent
halfopen intervals, but transparently behave as Ruby Floats.

## Setup

To install, type

```bash
sudo gem install greater_less
```

If you are using bundler, add `greater_less` to your gemfile

```ruby
gem 'greater_less'
```

## Getting Started

One easy way to integrate this gem into your project is by requiring the GreaterLess
string extension as follows:

```ruby
require 'greater_less/string_extension'
```

This extension redifines the `#to_f` method of the String class:

```ruby
class String
  alias :to_f_without_greater_less :to_f

  def to_f
    if self =~ GreaterLess::GREATER_LESS
      return GreaterLess.new(self)
    end
    self.to_f_without_greater_less
  end
end
```

Now when a string starts with a greater or less sign (like for instance
`"> 3.45"`), the `#to_f` method converts it to a GreaterLess object
instead of the value `0.0`.

Alternatively you can opt to create GreaterLess objects using `initialize` directly, like so:

```ruby
value = GreaterLess.new("> 3.45")
```

## Usage

A GreaterLess object can be compared to a Float (or other numeric) as if it were a
Float itself. For instance one can do the following:

```ruby
>> value = ">3.45".to_f
=> > 3.45
>> value > 2.45
=> true
>> value >= 2.45
=> true
>> 2.45 > value
=> false
>> 2.45 >= value
=> false
>> value == ">3.45".to_f
=> true
>> value != 2.45
=> true
```

It is also possible to compare GreaterLess objects with each other, so you
do not have to worry about what kind of object you are dealing with in your
code:

```ruby
>> value1 = ">3.45".to_f
=> > 3.45
>> value2 = "< 2.45".to_f
=> < 2.45
>> value1 > value2
=> true
>> value2 > value1
=> false
```

Finally it is possible to apply simple arithmetics to GreaterLess objects
like addition, subtraction, multiplication and division:

```ruby
>> value = ">3.45".to_f
=> > 3.45
>> value + 2
=> > 5.45
>> value - 2
=> > 1.4500000000000002
>> value * 2
=> > 1.725
```

Inverting the object's sign when multiplying with a negative numerical
or using a GreaterLess object in the denominator is nicely dealt with:

```ruby
>> value = ">3.45".to_f
=> > 3.45
>> -1 * value
=> < -3.45
>> 1 / value
=> < 0.2898550724637681
>> -1 / value
=> > -0.2898550724637681
```

In many cases it makes no sense to apply the operators +, -, * or / on 
a pair of GreaterLess objects, so when this happens an exception is 
raised for now.

All other methods are delegated to the Float value the GreaterLess
object contains, so that it transparently acts like a Float. This means you should be careful when using methods like
`abs` or `round` since they return the result from the underlying Float and bypass the GreaterLess behavior.

## Contributing to greater_less
 
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Copyright

Copyright (c) 2016 Samuel Esposito, Jorn van de Beek. See LICENSE.txt for further details.


CHANGELOG.md
### 1.0.0 / 2016-09-28

* Dropped support for ruby < 2.2


test/helper.rb
require 'rubygems'
require 'bundler'
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end
require 'test/unit'
require 'shoulda'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'greater_less'

class Test::Unit::TestCase
end


spec/spec_helper.rb
require 'rubygems'
if ENV['COVERAGE']
  require 'simplecov'
  SimpleCov.start
end

# --- Instructions ---
# - Sort through your spec_helper file. Place as much environment loading
#   code that you don't normally modify during development in the
#   Spork.prefork block.
# - Place the rest under Spork.each_run block
# - Any code that is left outside of the blocks will be ran during preforking
#   and during each_run!
# - These instructions should self-destruct in 10 seconds.  If they don't,
#   feel free to delete them.
#




$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rspec'
require 'andand'
require 'greater_less'

# Requires supporting files with custom matchers and macros, etc,
# in ./support/ and its subdirectories.
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}

RSpec.configure do |config|

end


spec/greater_less_spec.rb
require 'spec_helper'

describe GreaterLess do

  context "when the object is initialized with a string" do
    context "and the string starts with a greater than sign" do
      subject { GreaterLess.new("> 4.5") }

      it "should differ from its float value" do
        expect(subject == 4.5).to eq false
        expect(subject != 4.5).to eq true
        expect(4.5 == subject).to eq false
        expect(4.5 != subject).to eq true
      end

      it "should be greater than its float value" do
        expect(subject >  4.5).to eq true
        expect(subject >= 4.5).to eq true
        expect(4.5 <  subject).to eq true
        expect(4.5 <= subject).to eq true
      end

      it "should be greater than any value that is less than its float value" do
        expect(subject >  4.49).to eq true
        expect(subject >= 4.49).to eq true
        expect(4.49 <  subject).to eq true
        expect(4.49 <= subject).to eq true
      end

      it "should not be greater than a value greater than its float value" do
        expect(subject >  4.51).to eq false
        expect(subject >= 4.51).to eq false
        expect(4.51 <  subject).to eq false
        expect(4.51 <= subject).to eq false
      end

      it "should not be less than a value that is greater than its float value" do
        expect(subject <  4.51).to eq false
        expect(subject <= 4.51).to eq false
        expect(4.51 >  subject).to eq false
        expect(4.51 >= subject).to eq false
      end

      it "should be greater than a GreaterLess object that has a smaller or equal value and a less sign" do
        expect(subject > GreaterLess.new("< 4.49")).to eq true
        expect(subject > GreaterLess.new("< 4.5") ).to eq true
      end

      it "should not be greater than a GreaterLess object that has a bigger value or a greater sign" do
        expect(subject > GreaterLess.new("< 4.51")).to eq false
        expect(subject > GreaterLess.new("> 4.49")).to eq false
      end
    end

    context "and the string starts with a less than sign" do
      subject { GreaterLess.new("<4.5") }

      it "should differ from its float value" do
        expect(subject == 4.5).to eq false
        expect(subject != 4.5).to eq true
        expect(4.5 == subject).to eq false
        expect(4.5 != subject).to eq true
      end

      it "should be less than its float value" do
        expect(subject <  4.5).to eq true
        expect(subject <= 4.5).to eq true
        expect(4.5 >  subject).to eq true
        expect(4.5 >= subject).to eq true
      end

      it "should be less than any value that is greater than its float value" do
        expect(subject <  4.51).to eq true
        expect(subject <= 4.51).to eq true
        expect(4.51 >  subject).to eq true
        expect(4.51 >= subject).to eq true
      end

      it "should not be less than a value less than its float value" do
        expect(subject <  4.49).to eq false
        expect(subject <= 4.49).to eq false
        expect(4.49 >  subject).to eq false
        expect(4.49 >= subject).to eq false
      end

      it "should not be greater than a value that is less than its float value" do
        expect(subject >  4.49).to eq false
        expect(subject >= 4.49).to eq false
        expect(4.49 <  subject).to eq false
        expect(4.49 <= subject).to eq false
      end
    end

    context "and the string contains no sign" do
      subject { GreaterLess.new("4.5") }

      it "should be a float" do
        expect(subject.class).to eq Float
      end
    end
  end

  describe ".initialize" do
    context "when it receives something different from a string or a numeric" do
      it "should raise an exception" do
        expect { GreaterLess.new(Object.new, true) }.to raise_error('Can\'t handle Object!')
      end
    end
  end

  context "when the object is initialized with a string containing a sign" do
    subject { GreaterLess.new(">4.5") }

    it "should be a float" do
      expect(subject).to be_a(Float)
    end

    it "should equal itself" do
      expect(subject == subject).to eq true
    end

    describe "#coerce" do
      it "should raise an exception if it is called on a GreaterLess object" do
        expect { subject.coerce(GreaterLess.new("<2.45")) }.to raise_error('Can\'t handle GreaterLess!')
      end
    end

    describe "#inverted_sign" do
      it "should return a less sign when the object's sign is a greater sign" do
        expect(GreaterLess.new("> 1").inverted_sign).to eq("<")
      end

      it "should return a greater sign when the object's sign is a less sign" do
        expect(GreaterLess.new("< 1").inverted_sign).to eq(">")
      end
    end

    describe "#*" do
      it "should return a GreaterLess object" do
        expect((subject * 4).class).to eq GreaterLess
        expect((4 * subject).class).to eq GreaterLess
      end

      it "should carry out multiplication on its float value" do
        expect((subject * 4).value).to eq 18
        expect((4 * subject).value).to eq 18
      end

      it "should retain the object's sign when the argument is a positive numeric" do
        expect((subject * 4).sign).to eq subject.sign
        expect((4 * subject).sign).to eq subject.sign
      end

      it "should invert the object's sign when the argument is a negative numeric" do
        expect((subject * -4).sign).to eq subject.inverted_sign
        expect((-4 * subject).sign).to eq subject.inverted_sign
      end

      it "should raise an exception when the argument is a GreaterLess object" do
        expect { subject * GreaterLess.new("<1.2") }.to raise_error('Can\'t handle GreaterLess!')
      end
    end

    describe "#/" do
      it "should return a GreaterLess object" do
        expect((subject / 4).class).to eq GreaterLess
        expect((4 / subject).class).to eq GreaterLess
      end

      it "should carry out division on its float value" do
        expect((subject / 4).value).to eq 4.5 / 4
        expect((4 / subject).value.to_s).to eq (4 / 4.5).to_s
      end

      it "should retain the object's sign when the denominator is a positive numeric" do
        expect((subject / 4).sign).to eq subject.sign
      end

      it "should retain the object's sign when the numerator is a negative numeric" do
        expect((-4 / subject).sign).to eq subject.sign
      end

      it "should invert the object's sign when the object itself is the denominator" do
        expect((4 / subject).sign).to eq subject.inverted_sign
      end

      it "should invert the object's sign when the denominator is a negative numeric" do
        expect((subject / -4).sign).to eq subject.inverted_sign
      end

      it "should raise an exception when the argument is a GreaterLess object" do
        expect { subject / GreaterLess.new("<1.2") }.to raise_error('Can\'t handle GreaterLess!')
      end
    end

    describe "#+" do
      it "should return a GreaterLess object" do
        expect((subject + 4).class).to eq GreaterLess
        expect((4 + subject).class).to eq GreaterLess
      end

      it "should carry out addition on its float value" do
        expect((subject + 4).value).to eq 8.5
        expect((4 + subject).value).to eq 8.5
      end

      it "should retain the object's sign" do
        expect((subject + 4).sign).to eq subject.sign
        expect((4 + subject).sign).to eq subject.sign
      end

      it "should raise an exception when the argument is a GreaterLess object" do
        expect { subject + GreaterLess.new("<1.2") }.to raise_error('Can\'t handle GreaterLess!')
      end
    end

    describe "#-" do
      it "should return a GreaterLess object" do
        expect((subject - 4).class).to eq GreaterLess
        expect((4 - subject).class).to eq GreaterLess
      end

      it "should carry out subtraction on its float value" do
        expect((subject - 4).value).to eq 0.5
        expect((4 - subject).value).to eq -0.5
      end

      it "should retain the object's sign when a numerical is subtracted from it" do
        expect((subject - 4).sign).to eq subject.sign
      end

      it "should invert the object's sign when it is subtracted form a numerical" do
        expect((4 - subject).sign).to eq subject.inverted_sign
      end

      it "should raise an exception when the argument is a GreaterLess object" do
        expect { subject - GreaterLess.new("<1.2") }.to raise_error('Can\'t handle GreaterLess!')
      end
    end

    describe "#-@" do
      it "should return a GreaterLess object" do
        expect((-subject).class).to eq GreaterLess
      end

      it "should negate the object's value" do
        expect((-subject).value).to eq -subject.value
      end

      it "should invert the object's sign" do
        expect((-subject).sign).to eq subject.inverted_sign
      end
    end

    describe "#to_f" do
      it "should return the object itself" do
        expect(subject.to_f.class).to eq GreaterLess
      end
    end

    describe "#to_s" do
      it "should include this sign" do
        expect(subject.to_s).to eq "> 4.5"
      end
    end

    describe "#inspect" do
      it "should include this sign" do
        expect(subject.inspect).to eq "> 4.5"
      end
    end

    describe "#is_a? #kind_of?" do
      it "should acknowledge the GreaterLess class" do
        expect(subject.is_a? GreaterLess).to eq true
        expect(subject.kind_of? GreaterLess).to eq true
        # be_a actually calls kind_of instead of is_a
        expect(subject).to be_a(GreaterLess)
      end

      it "should acknowledge the Float class" do
        expect(subject.is_a? Float).to eq true
        expect(subject.kind_of? Float).to eq true
        # be_a actually calls kind_of instead of is_a
        expect(subject).to be_a(Float)
      end
    end

    describe 'delegation' do
      subject { GreaterLess.new('>4.5') }
      it 'passes unknown methods on to the underlying float' do
        # having methods like round and abs be handled on the original float is very error prone
        # so we document this 'feature' here and warn people in the Readme
        expect(subject.round).to eq 5
      end

      it 'preserves blocks' do
        success = false
        subject.tap do success = true end
        expect(success).to eq true
      end

      it 'preserves blocks that also have arguments' do
        steps = []
        subject.step(7.5) do |i|
          steps << i
        end
        expect(steps).to eq [4.5, 5.5, 6.5, 7.5]
      end

      it 'does not break andand' do
        expect(subject.andand).to be_a(GreaterLess)
        expect(subject.andand).to eq(subject)
        expect(subject.andand.to_s).to eq('> 4.5')
      end
    end
  end
end


lib/greater_less.rb
# The GreaterLess class can be used to generate objects that represent
# halfopen intervals, but transparently behave as Ruby Floats. One easy way to
# integrate this class into your project is by requiring the GreaterLess
# string extension as follows:
#
#  require 'greater_less/string_extension'
#
# This extension redifines the +#to_f+ method of the String class:
#
#  class String
#    alias :to_f_without_greater_less :to_f
#
#    def to_f
#      if self =~ GreaterLess::GREATER_LESS
#        return GreaterLess.new(self)
#      end
#      self.to_f_without_greater_less
#    end
#  end
#
# Now when a string starts with a greater or less sign (like for instance
# <tt>"> 3.45"</tt>), the +#to_f+ method converts it to a GreaterLess object
# instead of the value +0.0+.
# 
# Of course you can opt to create GreaterLess objects using +initialize+ directly, like so:
# 
#  value = GreaterLess.new("> 3.45")
# 
# A GreaterLess object can be compared to a Float (or other numeric) as if it were a
# Float itself. For instance one can do the following:
#
#  >> value = ">3.45".to_f
#  => > 3.45
#  >> value > 2.45
#  => true
#  >> value >= 2.45
#  => true
#  >> 2.45 > value
#  => false
#  >> 2.45 >= value
#  => false
#  >> value == ">3.45".to_f
#  => true
#  >> value != 2.45
#  => true
#
# It is also possible to compare GreaterLess objects with each other, so you
# do not have to worry about what kind of object you are dealing with in your
# code:
#
#  >> value1 = ">3.45".to_f
#  => > 3.45
#  >> value2 = "< 2.45".to_f
#  => < 2.45
#  >> value1 > value2
#  => true
#  >> value2 > value1
#  => false
#
# Finally it is possible to apply simple arithmetics to GreaterLess objects
# like addition, subtraction, multiplication and division:
#
#  >> value = ">3.45".to_f
#  => > 3.45
#  >> value + 2
#  => > 5.45
#  >> value - 2
#  => > 1.4500000000000002
#  >> value * 2
#  => > 1.725
#
# Inverting the object's sign when multiplying with a negative numerical
# or using a GreaterLess object in the denominator is nicely dealt with:
#
#  >> value = ">3.45".to_f
#  => > 3.45
#  >> -1 * value
#  => < -3.45
#  >> 1 / value
#  => < 0.2898550724637681
#  >> -1 / value
#  => > -0.2898550724637681
#
# In many cases it makes no sense to apply the operators +, -, * or / on 
# a pair of GreaterLess objects, so when this happens an exception is 
# raised for now.
#
# All other methods are simply passed to the Float value the GreaterLess
# object contains, so that it transparently acts like a Float.
#
require 'delegate'

class GreaterLess < Delegator
  GREATER_LESS = /^[<>] ?/

  #:nodoc:
  class << self
    alias :old_new :new

    def new(content, coerce=false)
      if coerce or content =~ GREATER_LESS
        old_new(content)
      else
        content.to_f
      end
    end
  end

  def initialize(content)
    if content.is_a? String
      if content =~ /^>/
        @sign = ">"
      elsif content =~ /^</
        @sign = "<"
      end
      @float = content.gsub(/^[<>] ?/, "").to_f
    elsif content.is_a? Numeric
      @float = content.to_f
    else
      raise "Can't handle #{content.class}!"
    end
    super(@float)
  end

  def coerce(object)
    if object.is_a? Numeric and not object.is_a? self.class
      [GreaterLess.new(object, true), self]
    else
      raise "Can't handle #{object.class}!"
    end
  end

  # delegate any method not defined in this class to float
  def __getobj__
    @float
  end
  def __setobj__(obj)
    @float = obj
  end

  def sign
    @sign
  end

  def inverted_sign
    case @sign
    when ">"
      "<"
    when "<"
      ">"
    end
  end

  def value
    @float
  end

  #:doc:
  def ==(numerical)
    if numerical.is_a? self.class
      @sign == numerical.sign and @float == numerical.value
    else
      false
    end
  end

  def >(numerical)
    if numerical.is_a? self.class
      @float >= numerical.value and [nil, ">"].include? @sign and [nil, "<"].include? numerical.sign
    else
      @float >= numerical and @sign == ">"
    end
  end

  def <(numerical)
    numerical > self
  end

  define_method("!=") do |numerical|
    not (self == numerical)
  end

  def >=(numerical)
    self == numerical or self > numerical
  end

  def <=(numerical)
    self == numerical or self < numerical
  end

  def *(numerical)
    value, sign = if numerical.is_a? self.class
                    raise "Can't handle #{self.class}!" if @sign
                    [@float * numerical.value, @float > 0 ? numerical.sign : numerical.inverted_sign]
                  else
                    [@float * numerical, numerical > 0 ? @sign : inverted_sign]
                  end
    GreaterLess.new("#{sign} #{value}")
  end

  def /(numerical)
    value, sign = if numerical.is_a? self.class
                    raise "Can't handle #{self.class}!" if @sign
                    [@float / numerical.value, @float > 0 ? numerical.inverted_sign : numerical.sign]
                  else
                    [@float / numerical, numerical > 0 ? @sign : inverted_sign]
                  end
    GreaterLess.new("#{sign} #{value}")
  end

  def +(numerical)
    value, sign = if numerical.is_a? self.class
                    raise "Can't handle #{self.class}!" if @sign
                    [@float + numerical.value, numerical.sign]
                  else
                    [@float + numerical, @sign]
                  end
    GreaterLess.new("#{sign} #{value}")
  end

  def -(numerical)
    self + -numerical
  end

  def -@
    GreaterLess.new("#{inverted_sign} -#{value}")
  end

  #:nodoc:
  def to_f
    self
  end

  def to_s
    "#{@sign} #{@float}"
  end

  def inspect
    self.to_s
  end

  def kind_of?(klass)
    is_a?(klass)
  end

  def is_a?(klass)
    if klass == self.class
      true
    else
      @float.is_a? klass
    end
  end

  # fixes delegator breaking andand functionality
  def andand(*)
    self
  end
end
